Ignore EPIPE from pager process

samumbach opened the pull request:
See also denysdovhan/bash-handbook#51 and denysdovhan/bash-handbook#46.

While this change achieves the goal of ignoring EPIPE (caused by quitting the pager; not a cause for concern), it makes handling any other errors more difficult. I'm not sure how best to do this with node.js.

Ideally, we could detect and ignore EPIPE then handle all other errors in the same way as `EventEmitter.prototype.emit`: call 'error' event handlers if present, throw the Error otherwise. This may be an appropriate use for [Domain](https://nodejs.org/api/domain.html), but this module is deprecated and I'm not aware of a replacement.

samumbach committed:
Ignore EPIPE from pager process

eush77 commented:
Thank you for contribution!

However, according to denysdovhan/bash-handbook#46 the problem is missing `error` handler — and the handler can be anything. (Empty handler works, too.) I think we shouldn't suppress all errors, but maybe reemit an error if it isn't EPIPE?

Also, should we be aiming at ECONNRESET, too?

```
$ seq 1000000 |node -e 'process.stdin.pipe(require("default-pager")())'
events.js:141
      throw er; // Unhandled 'error' event
      ^

Error: read ECONNRESET
    at exports._errnoException (util.js:860:11)
    at Pipe.onread (net.js:544:26)
```

And lastly, I fail to reproduce denysdovhan/bash-handbook#46, do you have a minimal example? This would be helpful.

eush77 commented on index.js:
```diff
@@ -20,6 +20,10 @@ module.exports = function (opts, cb) {
     });
 
+    ps.stdin.on('error', function (e) {
+        if (!e.errno || e.errno !== 'EPIPE') { throw(e); }
```
`throw` doesn't feel like a proper way to re-throw an asynchronous error…

sumbach replied:
I am completely at a loss as to how to handle this sanely. [events.js](https://github.com/nodejs/node/blob/dccccbb/lib/events.js#L155) throws an exception when no error handler is defined, otherwise it calls all error handlers in order--I'd like the ability to detect EPIPE (and possibly other errors) and ignore it without calling other error handlers, but provide default behavior for other errors (calling registered error handlers or throw if none present). But I can't figure out a way to do this within the framework of `EventEmitter`.

eush77 replied:
Yeah, I wonder what's the best we could do here.

1. We can't re-emit an error event because we are not able to suppress it in the first place.
2. Yet handling error event is the only way to ignore EPIPE and ECONNRESET.
3. These errors are not in any sense fatal, they're just distraction.

So unless you come up with something better I suggest to just add an empty handler with a comment for future reference.

```js
ps.stdin.on('error', function () {
    // Ignore EPIPE and ECONNRESET.                                             
});
```

sumbach replied:
Agreed. I'll update the PR now.

sumbach commented:
> do you have a minimal example?

Unfortunately, I was unable to run your `seq` example as you posted above (with node.js v5.0.0 and v4.2.6):

```
$ seq 1000000 | node -e 'process.stdin.pipe(require("default-pager")())'

(node) tty.setRawMode is deprecated. Use process.stdin.setRawMode instead.
tty.js:20
    throw new Error('can\'t set raw mode on non-tty');
    ^

Error: can't set raw mode on non-tty
    at Object.<anonymous> (tty.js:20:11)
    at Object.deprecated (internal/util.js:52:15)
    at setRaw (/Users/samumbach/src/bash-handbook/node_modules/default-pager/index.js:28:69)
    at module.exports (/Users/samumbach/src/bash-handbook/node_modules/default-pager/index.js:14:5)
    at [eval]:1:44
    at Object.exports.runInThisContext (vm.js:54:17)
    at Object.<anonymous> ([eval]-wrapper:6:22)
    at Module._compile (module.js:425:26)
    at node.js:587:27
    at doNTCallback0 (node.js:428:9)
```

But I was able to modify this into an example demonstrating the EPIPE error:

```
$ seq 1000000 > f.txt
$ node -e 'require("fs").createReadStream(require("path").join(__dirname, "./f.txt")).pipe(require("default-pager")())'
```

If I press `q` to exit your pager before scrolling to the bottom of the file, I receive the following:

```
events.js:141
      throw er; // Unhandled 'error' event
      ^

Error: write EPIPE
    at Object.exports._errnoException (util.js:860:11)
    at exports._exceptionWithHostPort (util.js:883:20)
    at WriteWrap.afterWrite (net.js:763:14)
```

eush77 commented:
> I was unable to run your `seq` example as you posted above

I'm sorry I forgot to push the fix for this. dfe81c9554f1c3092d905ef8f708d514a5a2c20b

> I was able to modify this into an example demonstrating the EPIPE error

And I still get ECONNRESET instead. I believe this is a platform issue.

```js
events.js:141
      throw er; // Unhandled 'error' event
      ^

Error: read ECONNRESET
    at exports._errnoException (util.js:856:11)
    at Pipe.onread (net.js:544:26)
```

sumbach commented:
> Upd: actually, I am able to get EPIPE if I exit the pager fast enough.

That's interesting... I assumed it was just a platform difference, but there seems to be a timing aspect as well.

There's definitely *some* connection between EPIPE and ECONNRESET, though I've had trouble getting a definitive answer on the subject:
- https://github.com/nodejs/node/blob/6fff47f/test/pummel/test-https-ci-reneg-attack.js#L82-L83
- https://github.com/nodejs/node/blob/6fff47f/test/pummel/test-tls-ci-reneg-attack.js#L80-L81
- http://stackoverflow.com/questions/2974021/what-does-econnreset-mean-in-the-context-of-an-af-local-socket/2979806#2979806
- http://vzkernel.blogspot.com/2012/09/connection-reset-vs-broken-pipe.html
- https://github.com/andris9/rai/issues/1#issuecomment-53087201

From all that, I wouldn't expect to see ECONNRESET from anything other than a TCP socket...

samumbach committed:
Ignore all errors from pager process' stdin file handle

- When the pager process is terminated before consuming all output from the
  parent JavaScript process, an EPIPE or ECONNRESET error is emitted
  (the particular error is platform- and timing-dependent).
- This is an expected behavior and doesn't indicate a failure; ideally, we would
  ignore these two errors so users don't get a confusing error message.
- But there's no clear way to ignore these two errors while retaining default
  behavior for others; instead, we've chosen to ignore all errors.

eush77 commented:
Merged in 8d4552a363e27968f1c9d8e92a8085688c52ccae and released in 1.1.0. Thanks!

samumbach commented:
@eush77 Just tested bash-handbook with default-pager 1.1.0 -- problem solved!! Thanks! :grinning: :sparkles:

eush77 commented:
:tada:
